File: src/viberoom/engine/source.py
"""Source frames for the client-side (GPU) preview.

The browser renders the pointwise part of the pipeline itself so that dragging
a slider is not a server round-trip. For that it needs the *decoded linear*
frame — the very array `render_preview` hands to `render()` — with its
highlight headroom intact.

PNG is not an option. Browsers decode PNG through ImageBitmap/drawImage, which
truncates every channel to 8 bits, and there is no way to get a 16-bit PNG's
samples into a WebGL texture without shipping a decoder. So the wire format is
raw texture bytes in one of the two layouts WebGL2 uploads directly:
RGB9_E5 (4 bytes/px, the default) and RGBA16F (8 bytes/px, the fallback).

The frame is produced at exactly the resolution `render_preview` renders at,
not at the resolution of the JPEG it finally emits. Anything else and the
client's frame and the server's frame are different images, so the swap from
the GPU preview to the settled server render would visibly jump.
"""

from __future__ import annotations

import hashlib
import struct
from pathlib import Path

import numpy as np

from viberoom.engine.cache import (
    PIPELINE_VERSION,
    _decode_cache,
    _downscale_linear,
    _preview_scale,
)
from viberoom.recipe.schema import Recipe

#: Wire formats the client may ask for. RGB9_E5 halves the payload and is
#: enough for a preview; RGBA16F exists for drivers that mishandle the packed
#: format, which is not hypothetical on older mobile GL.
SOURCE_FORMATS = ("rgb9e5", "rgba16f")

# GL_UNSIGNED_INT_5_9_9_9_REV: three 9-bit mantissas sharing one 5-bit
# exponent, biased by 15. Largest representable value is (2^9-1)/2^9 * 2^16.
_MANTISSA_BITS = 9
_EXP_BIAS = 15
_MAX_EXP = 31
_RGB9E5_MAX = 65408.0

#: float16 saturates here; beyond it the cast produces inf, which a shader
#: then propagates through every later op as NaN.
_F16_MAX = 65504.0

#: The cache file carries its own dimensions so a warm hit does not have to
#: decode the image again just to answer "how big is it".
_HEADER = struct.Struct("<II")


def encode_rgb9e5(arr: np.ndarray) -> bytes:
    """Pack float RGB into GL_UNSIGNED_INT_5_9_9_9_REV, 4 bytes per pixel.

    The shared exponent is why this format is here at all: it keeps relative
    precision across the whole frame, so linear values well above 1.0 — the
    highlight headroom a RAW decode carries — survive the trip, which no
    8-bit-per-channel transport can offer.
    """
    x = np.nan_to_num(np.asarray(arr, dtype=np.float32)[..., :3], nan=0.0)
    x = np.clip(x, 0.0, _RGB9E5_MAX)
    maxc = x.max(axis=-1)

    # exponent = floor(log2(max)) + 1 + bias. The all-zero pixel has no
    # logarithm, so it is substituted out and lands on a zero mantissa anyway.
    exp = np.floor(np.log2(np.where(maxc > 0, maxc, 1.0)))
    exp = np.maximum(exp, -_EXP_BIAS - 1) + 1 + _EXP_BIAS

    # Rounding the shared maximum can carry out of nine bits (511.5 -> 512);
    # the spec's remedy is to bump the exponent and quantize again.
    quantum = np.exp2(exp - _EXP_BIAS - _MANTISSA_BITS)
    exp = np.where(np.floor(maxc / quantum + 0.5) >= 512, exp + 1, exp)
    exp = np.clip(exp, 0, _MAX_EXP)

    quantum = np.exp2(exp - _EXP_BIAS - _MANTISSA_BITS)[..., None]
    m = np.clip(np.floor(x / quantum + 0.5), 0, 511).astype(np.uint32)

    packed = (
        m[..., 0]
        | (m[..., 1] << _MANTISSA_BITS)
        | (m[..., 2] << (2 * _MANTISSA_BITS))
        | (exp.astype(np.uint32) << (3 * _MANTISSA_BITS))
    )
    return packed.astype("<u4").tobytes()


def encode_rgba16f(arr: np.ndarray) -> bytes:
    """Half-float RGBA, 8 bytes per pixel, alpha pinned to 1.0."""
    x = np.clip(np.nan_to_num(np.asarray(arr, dtype=np.float32)[..., :3], nan=0.0), 0.0, _F16_MAX)
    out = np.empty(x.shape[:2] + (4,), dtype="<f2")
    out[..., :3] = x
    out[..., 3] = 1.0
    return out.tobytes()


def source_cache_key(path: Path, size: int, fmt: str) -> str:
    # PIPELINE_VERSION is in here because the *resolution* this frame is cut
    # to is a pipeline decision (`_preview_scale`), not just a decode one.
    raw = f"src-v{PIPELINE_VERSION}|{path}|{path.stat().st_mtime_ns}|{size}|{fmt}"
    return hashlib.sha1(raw.encode()).hexdigest()


def cached_source_dims(
    path: Path, size: int, fmt: str, cache_dir: Path
) -> tuple[int, int] | None:
    """Dimensions of an already-encoded frame, or None if it isn't cached.

    Lets a conditional request be answered with the full set of X-Source-*
    headers without paying for a decode.
    """
    cached = cache_dir / f"{source_cache_key(path, size, fmt)}.src"
    if not cached.exists():
        return None
    with cached.open("rb") as fh:
        head = fh.read(_HEADER.size)
    if len(head) < _HEADER.size:
        return None
    w, h = _HEADER.unpack(head)
    return w, h


def source_bytes(path: Path, size: int, fmt: str, cache_dir: Path) -> tuple[bytes, int, int]:
    """Encoded linear frame plus the dimensions it was encoded at, disk-cached."""
    if fmt not in SOURCE_FORMATS:
        raise ValueError(f"unknown source format {fmt!r}")

    cached = cache_dir / f"{source_cache_key(path, size, fmt)}.src"
    if cached.exists():
        blob = cached.read_bytes()
        w, h = _HEADER.unpack_from(blob)
        return blob[_HEADER.size :], w, h

    linear = _decode_cache.get(path, half_size=True)
    # A default recipe, deliberately: the client renderer only ever handles
    # recipes with no crop, so the scale it needs is the uncropped one — and
    # keying the cache on the recipe would defeat the whole point of a frame
    # that outlives every slider move.
    frame = _downscale_linear(linear, _preview_scale(linear, Recipe(), size))
    data = encode_rgb9e5(frame) if fmt == "rgb9e5" else encode_rgba16f(frame)

    h, w = frame.shape[:2]
    cached.write_bytes(_HEADER.pack(w, h) + data)
    return data, w, h

File: frontend/src/local/decode/source.ts
/** Encoding a decoded linear frame into the texture bytes the GPU renderer
 *  uploads — the client-side replacement for `engine/source.py`.
 *
 *  The wire formats are unchanged, and they have to be: `gpu/renderer.ts`
 *  uploads these buffers straight to WebGL2, so the layout is fixed by the
 *  driver, not by us. RGB9_E5 is 4 bytes/px and the default; RGBA16F is
 *  8 bytes/px for drivers that mishandle the packed format.
 *
 *  There is no PNG option for the same reason there never was: browsers decode
 *  PNG through ImageBitmap, which truncates every channel to 8 bits and throws
 *  away exactly the highlight headroom a RAW decode exists to carry.
 */

import type { LinearImage } from './types.ts'

export type SourceFormat = 'rgb9e5' | 'rgba16f'

/** Structurally identical to `gpu/source.ts`'s SourceFrame, so this drops into
 *  the place the `/source` fetch used to occupy. Redeclared rather than
 *  imported to keep the decode layer free of a dependency on the renderer. */
export interface SourceFrame {
  width: number
  height: number
  format: SourceFormat
  data: ArrayBuffer
}

// GL_UNSIGNED_INT_5_9_9_9_REV: three 9-bit mantissas sharing one 5-bit
// exponent, biased by 15. Largest representable value is (2^9-1)/2^9 * 2^16.
const MANTISSA_BITS = 9
const EXP_BIAS = 15
const MAX_EXP = 31
const RGB9E5_MAX = 65408

/** float16 saturates here; beyond it the cast produces inf, which a shader
 *  then propagates through every later op as NaN. */
const F16_MAX = 65504

function clamp(x: number, hi: number): number {
  if (!(x > 0)) return 0 // also catches NaN
  return x > hi ? hi : x
}

/** Packs linear RGB into GL_UNSIGNED_INT_5_9_9_9_REV, 4 bytes per pixel. */
export function encodeRgb9e5(img: LinearImage): ArrayBuffer {
  const src = img.data
  const out = new Uint32Array(img.width * img.height)
  for (let i = 0, p = 0; i < out.length; i++, p += 3) {
    const r = clamp(src[p], RGB9E5_MAX)
    const g = clamp(src[p + 1], RGB9E5_MAX)
    const b = clamp(src[p + 2], RGB9E5_MAX)
    const maxc = Math.max(r, g, b)

    // The all-zero pixel has no logarithm, so it is substituted out and lands
    // on a zero mantissa anyway.
    let exp = Math.max(Math.floor(Math.log2(maxc > 0 ? maxc : 1)), -EXP_BIAS - 1) + 1 + EXP_BIAS
    // Rounding the shared maximum can carry out of nine bits (511.5 -> 512);
    // the spec's remedy is to bump the exponent and quantize again.
    if (Math.floor(maxc / 2 ** (exp - EXP_BIAS - MANTISSA_BITS) + 0.5) >= 512) exp += 1
    exp = Math.min(Math.max(exp, 0), MAX_EXP)

    const quantum = 2 ** (exp - EXP_BIAS - MANTISSA_BITS)
    const mr = Math.min(511, Math.floor(r / quantum + 0.5))
    const mg = Math.min(511, Math.floor(g / quantum + 0.5))
    const mb = Math.min(511, Math.floor(b / quantum + 0.5))
    out[i] = (mr | (mg << MANTISSA_BITS) | (mb << (2 * MANTISSA_BITS)) | (exp << (3 * MANTISSA_BITS))) >>> 0
  }
  return out.buffer
}

/** Half-float RGBA, 8 bytes per pixel, alpha pinned to 1.0. */
export function encodeRgba16f(img: LinearImage): ArrayBuffer {
  const src = img.data
  const n = img.width * img.height
  // Float16Array is too new to rely on, so the conversion goes through a
  // Float32Array view of a scratch buffer and DataView's rounding.
  const out = new Uint16Array(n * 4)
  const f32 = new Float32Array(1)
  const bits = new Uint32Array(f32.buffer)

  /** Round-to-nearest-even, as both numpy's float16 cast and the GPU do. Plain
   *  truncation would bias every channel down by half a quantum, which at the
   *  bottom of the range is a visible lift lost from the shadows. */
  const half = (x: number): number => {
    const v = clamp(x, F16_MAX)
    if (v < 6.103515625e-5) {
      // Subnormal: quantum is a fixed 2^-24, so this is one rounded division.
      const m = v * 16777216
      const down = Math.floor(m)
      const frac = m - down
      return down + (frac > 0.5 || (frac === 0.5 && (down & 1)) ? 1 : 0)
    }
    f32[0] = v
    const b = bits[0]
    // Carrying out of the mantissa lands in the exponent field on its own,
    // which is why exponent and mantissa are incremented as one number.
    let packed = ((((b >>> 23) & 0xff) - 112) << 10) | ((b >>> 13) & 0x3ff)
    const rest = b & 0x1fff
    if (rest > 0x1000 || (rest === 0x1000 && (packed & 1))) packed++
    return packed
  }
  const ONE = half(1)
  for (let i = 0, p = 0, o = 0; i < n; i++, p += 3, o += 4) {
    out[o] = half(src[p])
    out[o + 1] = half(src[p + 1])
    out[o + 2] = half(src[p + 2])
    out[o + 3] = ONE
  }
  return out.buffer
}

export function encodeSource(img: LinearImage, format: SourceFormat): SourceFrame {
  return {
    width: img.width,
    height: img.height,
    format,
    data: format === 'rgb9e5' ? encodeRgb9e5(img) : encodeRgba16f(img),
  }
}
